File: webview_console/bridge.py
from __future__ import annotations

from typing import Any


class Bridge:
    def __init__(self, runtime) -> None:
        self._runtime = runtime

    def get_settings(self) -> dict[str, Any]:
        return self._runtime.get_settings()

    def update_settings(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self._runtime.update_settings(payload)

    def get_runtime(self) -> dict[str, Any]:
        return self._runtime.get_runtime()

    def get_pending_events(self, payload: dict[str, Any] | None = None) -> dict[str, Any]:
        return self._runtime.get_pending_events(payload or {})

    def create_run(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self._runtime.create_run(payload)

    def get_run(self, run_id: str) -> dict[str, Any]:
        return self._runtime.get_run(run_id)

    def cancel_run(self, run_id: str) -> dict[str, Any]:
        return self._runtime.cancel_run(run_id)

    def pause_run(self, run_id: str) -> dict[str, Any]:
        return self._runtime.pause_run(run_id)

    def resume_run(self, run_id: str) -> dict[str, Any]:
        return self._runtime.resume_run(run_id)

    def get_run_stages(self, run_id: str) -> dict[str, Any]:
        return self._runtime.get_run_stages(run_id)

    def get_run_logs(self, run_id: str) -> dict[str, Any]:
        return self._runtime.get_run_logs(run_id)

    def open_path(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self._runtime.open_path(payload)

    def select_directory(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self._runtime.select_directory(payload)

    def get_about(self) -> dict[str, Any]:
        return self._runtime.get_about()

    def check_update(self) -> dict[str, Any]:
        return self._runtime.check_update()

    def get_incremental_status(self, payload: dict[str, Any] | None = None) -> dict[str, Any]:
        return self._runtime.get_incremental_status(payload or {})

    def get_visualization_index(self, payload: dict[str, Any] | None = None) -> dict[str, Any]:
        return self._runtime.get_visualization_index(payload or {})

    def get_run_history(self) -> dict[str, Any]:
        return self._runtime.get_run_history()

    def export_run_history(self) -> dict[str, Any]:
        return self._runtime.export_run_history()

    def confirm_close_window(self) -> dict[str, Any]:
        return self._runtime.confirm_close_window()

    def cancel_close_window_request(self) -> dict[str, Any]:
        return self._runtime.cancel_close_window_request()

File: webui/src/services/bridge.ts
import type { ApiResult } from "@/services/types";
import type {
  HistorySnapshotResponse,
  LogSnapshotResponse,
  RunSnapshot,
  RuntimeSnapshot,
  StageSnapshotResponse,
  UpdateCheckResult,
  VisualizationIndexSnapshot,
} from "@/services/types";
import { BridgeError, normalizeBridgeError } from "@/services/errors";

const REQUIRED_METHODS = [
  "get_settings",
  "update_settings",
  "get_runtime",
  "get_pending_events",
  "create_run",
  "get_run",
  "cancel_run",
  "pause_run",
  "resume_run",
  "get_run_stages",
  "get_run_logs",
  "open_path",
  "select_directory",
  "get_about",
  "check_update",
  "get_incremental_status",
  "get_visualization_index",
  "get_run_history",
  "export_run_history",
  "confirm_close_window",
  "cancel_close_window_request",
] as const;

function hasRequiredMethods(candidate: Record<string, any> | undefined): boolean {
  if (!candidate) return false;
  return REQUIRED_METHODS.every((method) => typeof candidate[method] === "function");
}

function currentApi() {
  return window.pywebview?.api as Record<string, any> | undefined;
}

let readyPromise: Promise<void> | null = null;

function resetReadyPromise() {
  readyPromise = null;
}

function createReadyPromise(timeoutMs: number, intervalMs: number) {
  return new Promise<void>((resolve, reject) => {
    const startedAt = Date.now();
    let settled = false;
    let timerId: number | null = null;

    const finish = (error?: Error) => {
      if (settled) return;
      settled = true;
      window.removeEventListener("pywebviewready", onReady);
      if (timerId !== null) {
        window.clearInterval(timerId);
      }
      if (error) {
        resetReadyPromise();
        reject(error);
        return;
      }
      resolve();
    };

    const checkReady = () => {
      if (hasRequiredMethods(currentApi())) {
        finish();
        return true;
      }
      if (Date.now() - startedAt >= timeoutMs) {
        finish(new BridgeError("BRIDGE_NOT_READY"));
        return true;
      }
      return false;
    };

    const onReady = () => {
      window.setTimeout(() => {
        checkReady();
      }, 0);
    };

    window.addEventListener("pywebviewready", onReady, { once: false });

    if (checkReady()) {
      return;
    }

    timerId = window.setInterval(checkReady, intervalMs);
  });
}

async function waitForApi(timeoutMs = 15000, intervalMs = 100): Promise<void> {
  if (hasRequiredMethods(currentApi())) {
    return;
  }
  if (!readyPromise) {
    readyPromise = createReadyPromise(timeoutMs, intervalMs);
  }
  await readyPromise;
}

async function api() {
  await waitForApi();
  const bridge = currentApi();
  if (!bridge || !hasRequiredMethods(bridge)) {
    resetReadyPromise();
    throw new BridgeError("BRIDGE_NOT_READY");
  }
  return bridge;
}

async function call<T>(method: string, ...args: any[]): Promise<T> {
  try {
    const result = await (await api())[method](...args);
    const payload = result as ApiResult<T>;
    if (!payload.ok) {
      throw new BridgeError(payload.error?.code || "BRIDGE_CALL_FAILED", payload.error?.message);
    }
    return payload.data as T;
  } catch (error) {
    throw normalizeBridgeError(error, "BRIDGE_CALL_EXCEPTION", "桥接调用失败");
  }
}

export const bridge = {
  waitUntilReady: async (timeoutMs = 15000, intervalMs = 100) => {
    await waitForApi(timeoutMs, intervalMs);
  },
  getSettings: () => call("get_settings"),
  updateSettings: (payload: unknown) => call("update_settings", payload),
  getRuntime: () => call<RuntimeSnapshot>("get_runtime"),
  getPendingEvents: (limit = 200) => call<{ items: any[] }>("get_pending_events", { limit }),
  createRun: (payload: unknown) => call("create_run", payload),
  getRun: (runId: string) => call<RunSnapshot>("get_run", runId),
  cancelRun: (runId: string) => call("cancel_run", runId),
  pauseRun: (runId: string) => call("pause_run", runId),
  resumeRun: (runId: string) => call("resume_run", runId),
  getRunStages: (runId: string) => call<StageSnapshotResponse>("get_run_stages", runId),
  getRunLogs: (runId: string) => call<LogSnapshotResponse>("get_run_logs", runId),
  openPath: (path: string) => call("open_path", { path }),
  selectDirectory: (path: string) => call<{ path: string }>("select_directory", { path }),
  getAbout: () => call("get_about"),
  checkUpdate: () => call<UpdateCheckResult>("check_update"),
  getIncrementalStatus: (payload?: unknown) => call("get_incremental_status", payload || {}),
  getVisualizationIndex: (payload?: unknown) => call<VisualizationIndexSnapshot>("get_visualization_index", payload || {}),
  getRunHistory: () => call<HistorySnapshotResponse>("get_run_history"),
  exportRunHistory: () => call<{ path: string; count: number }>("export_run_history"),
  confirmCloseWindow: () => call("confirm_close_window"),
  cancelCloseWindowRequest: () => call("cancel_close_window_request"),
};
